servicetalk-concurrent-api/src/main/java/io/servicetalk/concurrent/api/DefaultCompositeCloseable.java

```
@@ -112,39 +125,85 @@ public void closeGracefully() {
-            return true;
-        }));
-        closeAsyncGracefully = closeAsyncGracefully.concat(closeable.closeAsyncGracefully().onErrorComplete(th -> {
-            //TODO: This should use concatDelayError when available.
-            LOGGER.debug("Ignored failure to close {}.", closeable, th);
-            return true;
-        }));
+        final Operand operand = getOrAddConcatOperand(true);
+        operand.closables.add(closeable);
+        resetState();
     }
 
     private void prependCloseableDelayError(final AsyncCloseable closeable) {
-        closeAsync = closeable.closeAsync().onErrorComplete(th -> {
-            //TODO: This should use prependDelayError when available.
-            LOGGER.debug("Ignored failure to close {}.", closeable, th);
-            return true;
-        }).concat(closeAsync);
-        closeAsyncGracefully = closeable.closeAsyncGracefully().onErrorComplete(th -> {
-            //TODO: This should use prependDelayError when available.
-            LOGGER.debug("Ignored failure to close {}.", closeable, th);
-            return true;
-        }).concat(closeAsyncGracefully);
+        final Operand operand = getOrAddConcatOperand(false);
+        operand.closables.add(closeable);
+        resetState();
+    }
+
+    private Completable buildCompletable(Function<AsyncCloseable, Completable> closerFunc) {
+        Completable result = completed();
+        for (Operand operand : operands) {
+            if (operand.isMerge) {
+                result = result.mergeDelayError(operand.closables.stream().map(closerFunc).toArray(Completable[]::new));
+            } else {
+                result = result.concat(operand.closables.stream().map(closerFunc).map(completable ->
+                        completable.onErrorComplete(th -> {
+                            // TODO: This should use concatDelayError when available.
+                            LOGGER.debug("Ignored failure to close", th);
+                            return true;
```

That was pre-existing, but looks like we hide all errors instead of delaying them. Should we try to save them in `AsyncContext` and verify its key at the end, as the last completable?

---

Lets defer for `concatDelayError` in a followup PR when it is available.